index.html

```diff
@@ -204,55 +204,1951 @@ <h3>
+            behavior for standard HID devices.
+          </div>
+        </dd>
+      </dl>
+      <p>
+        When an input report is received from a |device:HIDDevice|, run the
+        following steps:
+      </p>
+      <ol>
+        <li>If the input report is a [=blocked report=], abort these steps.
+        </li>
+        <li>Let |reportId:octet| be the report ID for this report, or 0 if
+        report IDs are not used.
+        </li>
+        <li>Let |data:DataView| be the data for this report, excluding the
+        report ID byte if report IDs are used.
+        </li>
+        <li>[=Fire an event=] named {{inputreport}} at |device| using
+        {{HIDInputReportEvent}} with its {{HIDInputReportEvent/device}}
+        attribute initialized to |device|, its {{HIDInputReportEvent/reportId}}
+        attribute initialized to |reportId|, and its
+        {{HIDInputReportEvent/data}} attribute initialized to |data|.
+        </li>
+      </ol>
+      <section>
+        <h3>
+          <dfn>open()</dfn> method
+        </h3>
+        <p>
+          The {{HIDDevice/open()}} method steps are:
+        </p>
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          </li>
+          <li>If [=this=].{{HIDDevice/[[state]]}} is not `"closed"`, [=reject=]
+          |promise| with a "{{InvalidStateError}}" {{DOMException}} and abort
+          these steps.
+          </li>
+          <li>Set [=this=].{{HIDDevice/[[state]]}} to `"opening"`.
+          </li>
+          <li>Perform the following steps [=in parallel=]:
+            <ol>
+              <li>Invoke the operating system to open the HID device.
+              </li>
+              <li>If this fails for any reason, [=queue a global task=] on the
+              [=relevant global object=] of [=this=] using the [=HID device
+              task source=] to [=reject=] |promise| with a
+              "{{NotAllowedError}}" {{DOMException}} and abort these steps.
+              </li>
+              <li>Set [=this=].{{HIDDevice/[[state]]}} to `"opened"`.
+              </li>
+              <li>[=Queue a global task=] on the [=relevant global object=] of
+              [=this=] using the [=HID device task source=] to [=resolve=]
+              |promise| with `undefined`.
+              </li>
+            </ol>
+          </li>
+          <li>Return |promise|.
+          </li>
+        </ol>
+      </section>
+      <section>
+        <h3>
+          <dfn>close()</dfn> method
+        </h3>
+        <p>
+          The {{HIDDevice/close()}} method steps are:
+        </p>
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          </li>
+          <li>Set [=this=].{{HIDDevice/[[state]]}} to `"closing"`.
+          </li>
+          <li>Perform the following steps [=in parallel=]:
+            <ol>
+              <li>Invoke the operating system to close the HID device and
+              release any associated resources.
+              </li>
```

Comment:
What happens to any unresolved `Promises` returned by the other methods on the `HIDDevice` interface?

Reply:
We should save a reference to any pending open, sendReport, sendFeatureReport, receiveFeatureReport promises so we can reject them when the device is closed. Is AbortError the right DOMException?

Reply:
If we actually abort the operation then yes. We should write some tests for this because the Chromium implementation probably actually blocks the resolution of `close()` on the completion of these methods because of the way it uses task runners for I/O operations. I think that's mostly fine, as `close()` is always going to race with the completion of these operations anyways.